File: java/src/test/java/com/callisto/sdk/ReporterTest.java
package com.callisto.sdk;

import com.callisto.sdk.errors.NetworkException;
import com.callisto.sdk.errors.ValidationException;
import com.callisto.sdk.reporting.ErrorReporter;
import com.callisto.sdk.reporting.ErrorSender;
import com.fasterxml.jackson.databind.JsonNode;
import org.junit.jupiter.api.Test;

import java.time.Duration;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

class ReporterTest {

    private static final String DSN = "https://app.callistosignal.com/ingest/uuid?key=abc123";

    /** Fake sender that records the last DSN + JSON and returns a configurable status. */
    static final class FakeSender implements ErrorSender {
        volatile String lastDsn;
        volatile String lastJson;
        final AtomicInteger calls = new AtomicInteger();
        int status = 202;
        boolean throwOnSend = false;

        @Override
        public int send(String dsn, String json) throws Exception {
            calls.incrementAndGet();
            this.lastDsn = dsn;
            this.lastJson = json;
            if (throwOnSend) {
                throw new RuntimeException("boom");
            }
            return status;
        }
    }

    private static CallistoClient client(StubHttpClient stub, FakeSender sender) {
        return new CallistoClient(
                "cid", "key", "https://api.example.com/v1", Duration.ofSeconds(5),
                stub, TestSupport.MAPPER,
                DSN, false, "production", sender);
    }

    private JsonNode payload(FakeSender sender) {
        assertNotNull(sender.lastJson, "expected a payload to have been sent");
        return TestSupport.json(sender.lastJson);
    }

    @Test
    void capturedApiErrorPostsToDsnWithMessageTypeLevel() {
        StubHttpClient stub = new StubHttpClient(404, "{\"message\": \"not found\"}");
        FakeSender sender = new FakeSender();
        try (CallistoClient client = client(stub, sender)) {
            assertThrows(com.callisto.sdk.errors.NotFoundException.class,
                    () -> client.balance().get());
            client.errorReporter().flush();

            assertEquals(DSN, sender.lastDsn);
            JsonNode p = payload(sender);
            assertEquals("not found", p.get("message").asText());
            assertEquals("com.callisto.sdk.errors.NotFoundException", p.get("type").asText());
            assertEquals("error", p.get("level").asText());
        }
    }

    @Test
    void contextSdkAndStatusCodeAndRequestPresentForTransportErrors() {
        StubHttpClient stub = new StubHttpClient(404, "{\"message\": \"not found\"}");
        FakeSender sender = new FakeSender();
        try (CallistoClient client = client(stub, sender)) {
            assertThrows(RuntimeException.class, () -> client.balance().get());
            client.errorReporter().flush();

            JsonNode p = payload(sender);
            JsonNode sdk = p.get("context").get("sdk");
            assertEquals(ErrorReporter.SDK_NAME, sdk.get("name").asText());
            assertEquals(ErrorReporter.SDK_LANGUAGE, sdk.get("language").asText());
            assertEquals(ErrorReporter.SDK_VERSION, sdk.get("version").asText());
            assertEquals("production", p.get("context").get("environment").asText());
            assertEquals(404, p.get("context").get("status_code").asInt());

            JsonNode request = p.get("request");
            assertEquals("GET", request.get("method").asText());
            assertEquals("/sms/balance", request.get("path").asText());
            assertEquals("GET /sms/balance", p.get("culprit").asText());
        }
    }

    @Test
    void networkErrorIsCaptured() {
        StubHttpClient stub = new StubHttpClient.Throwing();
        FakeSender sender = new FakeSender();
        try (CallistoClient client = client(stub, sender)) {
            assertThrows(NetworkException.class, () -> client.balance().get());
            client.errorReporter().flush();

            JsonNode p = payload(sender);
            assertEquals("com.callisto.sdk.errors.NetworkException", p.get("type").asText());
            assertEquals("GET", p.get("request").get("method").asText());
            assertEquals("/sms/balance", p.get("request").get("path").asText());
        }
    }

    @Test
    void noCredentialOrRequestBodyLeak() {
        StubHttpClient stub = new StubHttpClient(401, "{\"message\": \"invalid_key\"}");
        FakeSender sender = new FakeSender();
        try (CallistoClient client = client(stub, sender)) {
            assertThrows(RuntimeException.class,
                    () -> client.sms().send(com.callisto.sdk.resources.SmsSendRequest.builder()
                            .sender("Acme").to("+2250700000000").message("secret code 1234").build()));
            client.errorReporter().flush();

            String raw = sender.lastJson;
            assertNotNull(raw);
            // Credentials / auth must never appear.
            assertFalse(raw.contains("cid"), raw);
            assertFalse(raw.contains("\"key\""), raw);
            assertFalse(raw.toLowerCase().contains("authorization"), raw);
            assertFalse(raw.contains("Basic "), raw);
            // The outgoing request body (phone numbers, message content) must never leak.
            assertFalse(raw.contains("+2250700000000"), raw);
            assertFalse(raw.contains("secret code 1234"), raw);
        }
    }

    @Test
    void senderFailuresAreSwallowed() {
        StubHttpClient stub = new StubHttpClient(404, "{\"message\": \"nope\"}");
        FakeSender sender = new FakeSender();
        sender.throwOnSend = true;
        try (CallistoClient client = client(stub, sender)) {
            // Original error still propagates; capture path never throws.
            assertThrows(RuntimeException.class, () -> client.balance().get());
            client.errorReporter().flush();
            assertTrue(sender.calls.get() >= 1);
        }
    }

    @Test
    void nonAcceptedStatusIsSwallowed() {
        StubHttpClient stub = new StubHttpClient(404, "{\"message\": \"nope\"}");
        FakeSender sender = new FakeSender();
        sender.status = 401; // ingest rejects; reporter swallows
        try (CallistoClient client = client(stub, sender)) {
            assertThrows(RuntimeException.class, () -> client.balance().get());
            client.errorReporter().flush();
            assertEquals(1, sender.calls.get());
        }
    }

    @Test
    void captureNeverThrowsOnNull() {
        FakeSender sender = new FakeSender();
        ErrorReporter reporter = new ErrorReporter(DSN, null, sender, TestSupport.MAPPER);
        reporter.captureException(null);
        reporter.captureMessage(null);
        reporter.flush();
        reporter.close();
        assertEquals(0, sender.calls.get());
    }

    @Test
    void noOpWithoutDsn() {
        StubHttpClient stub = new StubHttpClient(404, "{\"message\": \"nope\"}");
        FakeSender sender = new FakeSender();
        // No DSN -> reporter disabled; behaves exactly as before.
        try (CallistoClient client = new CallistoClient(
                "cid", "key", "https://api.example.com/v1", Duration.ofSeconds(5),
                stub, TestSupport.MAPPER, null, false, null, sender)) {
            assertFalse(client.errorReporter().isEnabled());
            assertThrows(RuntimeException.class, () -> client.balance().get());
            client.errorReporter().flush();
            assertEquals(0, sender.calls.get());
            assertNull(sender.lastJson);
        }
    }

    @Test
    void invalidDsnIsNoOp() {
        FakeSender sender = new FakeSender();
        ErrorReporter reporter = new ErrorReporter("not a url", null, sender, TestSupport.MAPPER);
        assertFalse(reporter.isEnabled());
        reporter.captureMessage("hello", "info", null);
        reporter.flush();
        assertEquals(0, sender.calls.get());
    }

    @Test
    void publicCaptureMessageWorks() {
        FakeSender sender = new FakeSender();
        StubHttpClient stub = new StubHttpClient(200, "{}");
        try (CallistoClient client = client(stub, sender)) {
            client.captureMessage("something happened");
            client.errorReporter().flush();

            JsonNode p = payload(sender);
            assertEquals("something happened", p.get("message").asText());
            assertEquals("info", p.get("level").asText());
            assertNotNull(p.get("context").get("sdk"));
        }
    }

    @Test
    void publicCaptureExceptionWithLevelAndExtraAndUser() {
        FakeSender sender = new FakeSender();
        StubHttpClient stub = new StubHttpClient(200, "{}");
        try (CallistoClient client = client(stub, sender)) {
            client.setUser(Map.of("id", "u_1", "email", "user@example.com"));
            client.captureException(new IllegalStateException("bad state"),
                    "warning", Map.of("order_id", "ord_9"));
            client.errorReporter().flush();

            JsonNode p = payload(sender);
            assertEquals("bad state", p.get("message").asText());
            assertEquals("java.lang.IllegalStateException", p.get("type").asText());
            assertEquals("warning", p.get("level").asText());
            assertEquals("ord_9", p.get("context").get("order_id").asText());
            assertEquals("u_1", p.get("user").get("id").asText());
        }
    }

    @Test
    void clientSideValidationErrorIsCaptured() {
        FakeSender sender = new FakeSender();
        StubHttpClient stub = new StubHttpClient(200, "{}");
        try (CallistoClient client = client(stub, sender)) {
            assertThrows(ValidationException.class,
                    () -> client.notifications().send(
                            com.callisto.sdk.resources.NotifyRequest.builder().topic("welcome").build()));
            client.errorReporter().flush();

            JsonNode p = payload(sender);
            assertEquals("com.callisto.sdk.errors.ValidationException", p.get("type").asText());
        }
    }

    @Test
    void uncaughtHandlerInstallsAndChains() {
        Thread.UncaughtExceptionHandler original = Thread.getDefaultUncaughtExceptionHandler();
        AtomicInteger priorCalls = new AtomicInteger();
        Thread.setDefaultUncaughtExceptionHandler((t, e) -> priorCalls.incrementAndGet());
        FakeSender sender = new FakeSender();
        StubHttpClient stub = new StubHttpClient(200, "{}");
        try (CallistoClient client = new CallistoClient(
                "cid", "key", "https://api.example.com/v1", Duration.ofSeconds(5),
                stub, TestSupport.MAPPER, DSN, true, null, sender)) {
            Thread.UncaughtExceptionHandler installed = Thread.getDefaultUncaughtExceptionHandler();
            assertNotNull(installed);
            // Invoke directly (without crashing the test process); it should report + chain.
            installed.uncaughtException(Thread.currentThread(), new RuntimeException("fatal boom"));
            client.errorReporter().flush();

            JsonNode p = payload(sender);
            assertEquals("fatal", p.get("level").asText());
            assertEquals(1, priorCalls.get(), "pre-existing handler must be chained");
        } finally {
            Thread.setDefaultUncaughtExceptionHandler(original);
        }
    }

    @Test
    void rateLimitContextCarriesRetryAfter() {
        StubHttpClient stub = new StubHttpClient(429, "{\"message\": \"slow down\"}",
                Map.of("Retry-After", List.of("30")));
        FakeSender sender = new FakeSender();
        try (CallistoClient client = client(stub, sender)) {
            assertThrows(com.callisto.sdk.errors.RateLimitException.class,
                    () -> client.balance().get());
            client.errorReporter().flush();

            JsonNode p = payload(sender);
            assertEquals(429, p.get("context").get("status_code").asInt());
            assertEquals(30, p.get("context").get("retry_after").asInt());
        }
    }
}

File: go/reporter_test.go
package callisto

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"
	"strings"
	"testing"
)

const testDSN = "https://app.callistosignal.com/ingest/app-1?key=pub-key"

// newReportingClient returns a client with error reporting enabled against a
// fake sender, plus the API server handler.
func newReportingClient(t *testing.T, sender *fakeSender, handler http.HandlerFunc) *Client {
	t.Helper()
	t.Setenv("CALLISTO_ERROR_DSN", "")
	c, srv := newTestClient(t, handler)
	_ = srv
	rc, err := NewClient(Options{
		ClientID: "cid", APIKey: "secretkey", BaseURL: c.transport.cfg.baseURL,
		ErrorDSN: testDSN, ErrorSender: sender, Environment: "test",
	})
	if err != nil {
		t.Fatalf("NewClient: %v", err)
	}
	t.Cleanup(rc.Close)
	return rc
}

func TestReporterCapturesTransportError(t *testing.T) {
	sender := &fakeSender{ch: make(chan struct{}, 4)}
	c := newReportingClient(t, sender, func(w http.ResponseWriter, r *http.Request) {
		writeJSON(w, 404, `{"message":"not found"}`)
	})

	_, err := c.Balance.Get(context.Background(), BalanceParams{})
	// The original error must still propagate.
	var nf *NotFoundError
	if !errors.As(err, &nf) {
		t.Fatalf("call err = %v, want *NotFoundError", err)
	}

	waitForSends(t, sender, 1)
	dsn, payload := sender.last()
	if dsn != testDSN {
		t.Errorf("posted to %q, want %q", dsn, testDSN)
	}
	if payload["type"] != "NotFoundError" || payload["level"] != "error" {
		t.Errorf("payload type/level = %v/%v", payload["type"], payload["level"])
	}
	if payload["culprit"] != "GET /sms/balance" {
		t.Errorf("culprit = %v", payload["culprit"])
	}
	req, _ := payload["request"].(map[string]any)
	if req["method"] != "GET" || req["path"] != "/sms/balance" {
		t.Errorf("request = %v", payload["request"])
	}
	ctxMap, _ := payload["context"].(map[string]any)
	if ctxMap["status_code"] != 404 {
		t.Errorf("status_code = %v, want 404", ctxMap["status_code"])
	}
	sdk, _ := ctxMap["sdk"].(map[string]any)
	if sdk["language"] != "go" || sdk["name"] != sdkName {
		t.Errorf("sdk = %v", sdk)
	}
}

func TestReporterNoCredentialOrRequestBodyLeak(t *testing.T) {
	sender := &fakeSender{ch: make(chan struct{}, 4)}
	c := newReportingClient(t, sender, func(w http.ResponseWriter, r *http.Request) {
		writeJSON(w, 422, `{"message":"bad request"}`)
	})

	// This request body carries PII (phone + message) and credentials are on the
	// wire as Basic auth — none of it may appear in the reported payload.
	_, _ = c.SMS.Send(context.Background(), SendSMSParams{
		Sender: "Acme", To: []string{"+2250799999999"}, Message: "secret-message-text",
	})

	waitForSends(t, sender, 1)
	_, payload := sender.last()
	raw, err := json.Marshal(payload)
	if err != nil {
		t.Fatal(err)
	}
	blob := string(raw)
	for _, forbidden := range []string{
		"secretkey",           // API key
		"cid",                 // client ID
		"Authorization",       // auth header
		"+2250799999999",      // recipient PII
		"secret-message-text", // message content
	} {
		if strings.Contains(blob, forbidden) {
			t.Errorf("payload leaked %q: %s", forbidden, blob)
		}
	}
	// Sanity: the payload was actually built.
	if !strings.Contains(blob, "ValidationError") {
		t.Errorf("expected ValidationError type in payload: %s", blob)
	}
}

func TestReporterDisabledWithoutDSN(t *testing.T) {
	t.Setenv("CALLISTO_ERROR_DSN", "")
	sender := &fakeSender{}
	c, _ := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		writeJSON(w, 500, `{"message":"x"}`)
	})
	// Re-point the transport's reporter? Instead build a fresh client with the
	// sender but NO DSN, confirming it never sends.
	noDSN, err := NewClient(Options{
		ClientID: "cid", APIKey: "secretkey", BaseURL: c.transport.cfg.baseURL,
		ErrorSender: sender, // provided, but no DSN → disabled
	})
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(noDSN.Close)
	if noDSN.reporter.Enabled() {
		t.Fatal("reporter should be disabled without a DSN")
	}

	_, _ = noDSN.Balance.Get(context.Background(), BalanceParams{})
	noDSN.reporter.Flush()
	if sender.count() != 0 {
		t.Errorf("sender called %d times, want 0", sender.count())
	}
}

func TestReporterSwallowsSenderFailure(t *testing.T) {
	sender := &fakeSender{ch: make(chan struct{}, 4), err: errors.New("network down")}
	c := newReportingClient(t, sender, func(w http.ResponseWriter, r *http.Request) {
		writeJSON(w, 200, `{"credit":1,"currency":"X"}`)
	})

	// CaptureException must never panic even when the sender fails.
	c.CaptureException(errors.New("boom"), WithLevel("warning"))
	waitForSends(t, sender, 1)
	_, payload := sender.last()
	if payload["level"] != "warning" || payload["message"] != "boom" {
		t.Errorf("payload = %v", payload)
	}
}

func TestCaptureMessage(t *testing.T) {
	sender := &fakeSender{ch: make(chan struct{}, 4)}
	c := newReportingClient(t, sender, func(w http.ResponseWriter, r *http.Request) {})

	c.CaptureMessage("hello", "info")
	waitForSends(t, sender, 1)
	_, payload := sender.last()
	if payload["message"] != "hello" || payload["level"] != "info" {
		t.Errorf("payload = %v", payload)
	}
}

func TestCaptureExceptionConstrainsLevel(t *testing.T) {
	sender := &fakeSender{ch: make(chan struct{}, 4)}
	c := newReportingClient(t, sender, func(w http.ResponseWriter, r *http.Request) {})

	c.CaptureException(errors.New("x"), WithLevel("bogus"))
	waitForSends(t, sender, 1)
	_, payload := sender.last()
	if payload["level"] != "error" {
		t.Errorf("level = %v, want error (bogus constrained)", payload["level"])
	}
}

func TestSetUserAttachedToEvents(t *testing.T) {
	sender := &fakeSender{ch: make(chan struct{}, 4)}
	c := newReportingClient(t, sender, func(w http.ResponseWriter, r *http.Request) {})

	c.SetUser(map[string]any{"id": "u1"})
	c.CaptureMessage("m", "info")
	waitForSends(t, sender, 1)
	_, payload := sender.last()
	user, _ := payload["user"].(map[string]any)
	if user["id"] != "u1" {
		t.Errorf("user = %v", payload["user"])
	}
}
